### PR TITLE
Traceability extension protobufs

### PR DESCRIPTION
**Pull request by mustafauzunn**
**Description**:

Traceability extension protobufs

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: streams/sidecar_file.proto [hunk 1]
```diff
@@ -53,15 +55,43 @@ message TransactionSidecarRecord {
...
+
   /*
   * List of sidecar types.
   * In future there will be other categories.
   */
   oneof sidecar_records {
-    ContractStateChanges state_changes = 2;
+    ContractStateChanges state_changes = 3;
+    ContractActions actions = 4;
+    ContractBytecode bytecode = 5;
```
Comment (dimitar-dinev) [inline streams/sidecar_file.proto:67]: Q: Why do we not have a `ContractBytecodes` wrapper of `repeated ContractBytecode` as we have for `ContractStateChange` and `ContractAction`? How will we externalize all contract bytecodes when a `ContractCall/Create` creates > 1 contracts?

File: streams/sidecar_file.proto [hunk 2]
```diff
@@ -53,15 +55,43 @@ message TransactionSidecarRecord {
...
+
+  /**
+  * Sidecar generated from migration.
+  */
+  MIGRATION = 1;
+
+  /**
+  * Sidecar generated from trace.
+  */
+  TRACE = 2;
```
Comment (steven-sheehy) [inline streams/sidecar_file.proto:88]: I imagine in the future we might put other, non-traceability information in the sidecar like errata records so we should name this something more general. Or we can just change it to a `bool migration` which might be clearer.

Whichever data type, we should choose the default value to be the non-migration value so it doesn't incur the cost of writing extra bytes indefinitely to the file for the one-time migration. If we use `bool migration` then the default value is false and would not serialize the field for the majority of record files.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: streams/contract_action.proto [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    AccountID calling_account = 2;
+
+    /**
+     * If the caller was a smart contract account, the ContractID.
+     */
+    ContractID calling_contract = 3;
+  }
+
+  /**
+   * Gas given to this action, the upper limit of gas this action can spend.
```
Comment (Nana-EC) [inline streams/contract_action.proto:90]: nit: current wording makes it unclear which gas it refers to
```suggestion
   * The upper limit of gas this action can spend.
```

File: streams/contract_state_change.proto [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * The storage changes to a smart contract's storage as a side effect of the function call.
+ */
+message ContractStateChange {
+
+  /**
+   * The contract to which the storage changes apply to
+   */
+  ContractID contractID = 1;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply streams/contract_state_change.proto]: DONE

File: streams/record_stream_file.proto [hunk 5]
```diff
@@ -103,4 +103,6 @@ message SidecarMetadata {
 enum SidecarType {
   SIDECAR_TYPE_UNKNOWN = 0;
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply streams/record_stream_file.proto]: DONE

File: streams/contract_bytecode.proto [hunk 6]
```diff
@@ -0,0 +1,39 @@
+syntax = "proto3";
+
+package proto;
+
+/*-
+ * ‌
+ * Hedera Network Services Protobuf
+ * ​
+ * Copyright (C) 2018 - 2021 Hedera Hashgraph, LLC
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply streams/contract_bytecode.proto]: DONE

File: streams/contract_action.proto [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+  /**
+   * The actual gas spent by this action.
+   */
+  int64 gas_used = 10;
+
+  /**
+   * The result data of the action.
+   */
+  oneof resultData {
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply streams/contract_action.proto]: DONE

File: streams/sidecar_file.proto [hunk 8]
```diff
@@ -53,15 +55,43 @@ message TransactionSidecarRecord {
...
+}
+
 message ContractStateChanges {
   repeated ContractStateChange contract_state_changes = 1;
 }
+
+message ContractActions {
+  repeated ContractAction contract_actions = 1;
+}
```
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply streams/sidecar_file.proto]: DONE

File: streams/sidecar_file.proto (hunk 2 above)
Comment (mustafauzunn) [reply streams/sidecar_file.proto]: DONE

File: streams/contract_action.proto (hunk 3 above)
Comment (mustafauzunn) [reply streams/contract_action.proto]: DONE

**Review by steven-sheehy — DISMISSED** [review]
LGTM

File: streams/contract_state_change.proto [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+option java_package = "com.hederahashgraph.api.proto.java";
```
Comment (xin-hedera) [inline streams/contract_state_change.proto:25]: same here

**Review by xin-hedera — CHANGES_REQUESTED** [review]

File: streams/contract_action.proto [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * The type of action described by the action proto.
+ */
+enum ContractActionType {
+  /**
+   * Most CALL, CALLCODE, DELEGATECALL, and STATICCALL, and first action of ContractCall/ContractCallLocal to deployed
+   * contracts. This does not include calls to system or precompiled contracts.
+   */
+  CALL = 0;
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply streams/contract_action.proto:43]: What old protobuf? This is a new message for sidecar and it's a required field set by consensus nodes so there shouldn't be a need for an unknown value. I had him just remove it in the last commit.

File: streams/contract_bytecode.proto [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+option java_package = "com.hederahashgraph.api.proto.java";
```
[the remark this answers is not in the document]
Comment (steven-sheehy) [reply streams/contract_bytecode.proto:25]: Good catch. All files in `streams` should have a separate java package from services folder.

File: streams/contract_action.proto (hunk 10 above)
[the remark this answers is not in the document]
Comment (xin-hedera) [reply streams/contract_action.proto:43]: never mind. I thought if we add new fields to the enum in the future, the parser generated from the current version would parse new enum fields to `CALL`, it'll actually parse it to `ContractActionType.UNRECOGNIZED(-1)`.

File: streams/contract_bytecode.proto [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+  * Contract bytecode during deployment
+  */
+  bytes initcode = 1;
+
+  /**
+  * Contract bytecode after deployment
+  */
+  bytes runtime_bytecode = 2;
```
Comment (steven-sheehy) [inline streams/contract_bytecode.proto:38]: I think we would also need the `ContractID contract_id` here as well for migrations. For non-migration we can use the associated consensus timestamp to link it to the contract create. 

For migrations, the consensus timestamp is for a random transaction after startup and not associated with the contract create. So we would need the contract ID here or we would need the consensus timestamp to be linked to some contract create synthetic transaction.

File: streams/contract_bytecode.proto (hunk 11 above)
[the remark this answers is not in the document]
Comment (mustafauzunn) [reply streams/contract_bytecode.proto]: DONE

File: streams/contract_state_change.proto (hunk 9 above)
Comment (mustafauzunn) [reply streams/contract_state_change.proto]: DONE

**Review by xin-hedera — APPROVED** [review]
LGTM

**Review by steven-sheehy — DISMISSED** [review]
LGTM

File: streams/sidecar_file.proto (hunk 1 above)
Comment (stoqnkpL) [reply streams/sidecar_file.proto:70]: Yes, each contract deployment creates a separate record, e.g.
- deploying 5 contracts with a single `ContractCreate` grpc transaction would create the top-level creation record and 4 child creation records; each one will have a corresponding `ContractBytecode` sidecar record.
- deploying 5 contracts via a `ContractCall` grpc transaction would create the top-level call record and 5 creation records; each child will have a corresponding `ContractBytecode` sidecar record.

**Review by stoqnkpL — APPROVED** [review]
LGTM

**Review by stoqnkpL — CHANGES_REQUESTED** [review]
I think we should resolve this [comment](https://github.com/hashgraph/hedera-improvement-proposal/pull/513#discussion_r905277554) before merging.

File: streams/sidecar_file.proto (hunk 1 above)
Comment (dimitar-dinev) [reply streams/sidecar_file.proto:70]: Got it, thanks for the clarifications!

**Review by steven-sheehy — DISMISSED** [review]

**Review by xin-hedera — APPROVED** [review]

**Review by tinker-michaelj — DISMISSED** [review]
LGTM

**Comment by steven-sheehy** [comment]
Need to gpg sign all your commits

**Review by xin-hedera — APPROVED** [review]

**Pull request merged**